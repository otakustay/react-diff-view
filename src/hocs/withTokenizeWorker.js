import {wrapDisplayName} from 'recompose';
import {useTokenizeWorker} from '../hooks';

const defaultMapPayload = data => {
    return data;
};

export default (worker, options = {}) => {
    const {mapPayload = defaultMapPayload, ...hookOptions} = options;

    const resolveMessagePayload = props => {
        const {hunks, oldSource, language} = props;
        const input = {
            language: language,
            oldSource: oldSource,
            hunks: hunks,
        };
        return mapPayload(input, props);
    };

    return ComponentIn => {
        const ComponentOut = props => {
            const payload = resolveMessagePayload(props);
            const tokenizationResult = useTokenizeWorker(worker, payload, hookOptions);

            return <ComponentIn {...props} {...tokenizationResult} />;
        };

        ComponentOut.displayName = wrapDisplayName(ComponentIn, 'withTokenizeWorker');

        return ComponentOut;
    };
};
