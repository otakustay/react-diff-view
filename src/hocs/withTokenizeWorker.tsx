import {ComponentType} from 'react';
import {TokenizePayload, TokenizeResult, TokenizeWorkerOptions, useTokenizeWorker} from '../hooks';
import {HunkData, Source} from '../utils';
import {wrapDisplayName} from './wrapDisplayName';

export interface RequiredProps {
    hunks: HunkData[];
    oldSource: Source;
    language: string;
}

function defaultMapPayload(data: RequiredProps) {
    return data;
}

interface ToeknizeWorkerHocOptions<T extends TokenizePayload> extends TokenizeWorkerOptions<T> {
    mapPayload?: <P extends RequiredProps>(payload: RequiredProps, props: P) => any;
}

export default function withTokenizeWorkerwithTokenizeWorker<T extends TokenizePayload>(
    worker: Worker,
    options: ToeknizeWorkerHocOptions<T> = {}
) {
    const {mapPayload = defaultMapPayload, ...hookOptions} = options;

    function resolveMessagePayload<P extends RequiredProps>(props: P): T {
        const {hunks, oldSource, language} = props;
        const input = {language, oldSource, hunks};
        return mapPayload(input, props);
    }

    return function wrap<P extends RequiredProps>(ComponentIn: ComponentType<P>) {
        function ComponentOut(props: P & TokenizeResult) {
            const payload = resolveMessagePayload(props);
            const tokenizationResult = useTokenizeWorker(worker, payload, hookOptions);

            return <ComponentIn {...props} {...tokenizationResult} />;
        }

        ComponentOut.displayName = wrapDisplayName(ComponentIn, 'withTokenizeWorker');

        return ComponentOut;
    };
}
