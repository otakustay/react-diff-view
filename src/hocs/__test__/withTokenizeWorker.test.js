import renderer from 'react-test-renderer';
import withTokenizeWorker from '../withTokenizeWorker';

const journey = [];
let handleReceiveTokens = null;
const worker = {
    addEventListener(type, receiveTokens) {
        journey.push(['addEventListener', type, typeof receiveTokens]);
        handleReceiveTokens = receiveTokens;
    },
    removeEventListener(type, receiveTokens) {
        journey.push(['removeEventListener', type, typeof receiveTokens]);
        handleReceiveTokens = null;
    },
    postMessage(data) {
        journey.push(['postMessage', typeof data]);
        handleReceiveTokens({data});
    }
};
const ComponentIn = () => <div />;

describe('withTokenizeWorker', () => {
    test('returns a hoc function', () => {
        expect(typeof withTokenizeWorker(worker)).toBe('function');
    });

    test('hoc function return a component', () => {
        expect(typeof withTokenizeWorker(worker)(ComponentIn)).toBe('function');
    });

    test('render component', () => {
        const ComponentOut = withTokenizeWorker(worker)(ComponentIn);
        expect(renderer.create(<ComponentOut />).toJSON()).toMatchSnapshot();
        expect(journey).toEqual([
            ['addEventListener', 'message', 'function'],
            ['postMessage', 'object']
        ]);
    });
});
