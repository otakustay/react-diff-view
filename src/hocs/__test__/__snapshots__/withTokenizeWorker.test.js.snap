// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`withTokenizeWorker render component 1`] = `<div />`;
